homeassistant/components/pvpc_hourly_pricing/__init__.py

```diff
@@ -45,10 +66,67 @@ async def async_setup(hass: HomeAssistant, config: dict):
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up pvpc hourly pricing from a config entry."""
+    if len(entry.data) == 2:
+        defaults = {
+            ATTR_TARIFF: _DEFAULT_TARIFF,
+            ATTR_POWER: DEFAULT_POWER_KW,
+            ATTR_POWER_P3: DEFAULT_POWER_KW,
+        }
+        data = {**entry.data, **defaults}
+        hass.config_entries.async_update_entry(
+            entry, unique_id=_DEFAULT_TARIFF, data=data, options=defaults
+        )
+
+        @callback
+        def update_unique_id(reg_entry):
+            """Change unique id for sensor entity, pointing to new tariff."""
+            return {"new_unique_id": _DEFAULT_TARIFF}
+
+        try:
+            await async_migrate_entries(hass, entry.entry_id, update_unique_id)
+            _LOGGER.warning(
+                "Migrating PVPC sensor from old tariff '%s' to new '%s'. "
+                "Configure the integration to set your contracted power, "
+                "and select prices for Ceuta/Melilla, "
+                "if that is your case",
+                entry.data[ATTR_TARIFF],
+                _DEFAULT_TARIFF,
+            )
+        except ValueError:
```

Comment:
When does the ValueError happen?

Reply:
Hi @MartinHjelmare,

When there is >1 old sensors (very strange, as it has no practical application far from data monitoring, but it can be up to 3), in the conversion to the new tariff only the 1st survives, and for the others, because of the **new unique-id**, the entity_registry fails with that ValueError (printed here):

```
2021-06-15 08:30:13 WARNING (MainThread) [homeassistant.components.pvpc_hourly_pricing] Migrating PVPC sensor from old tariff 'discrimination' to new '2.0TD'. Configure the integration to set your contracted power, and select prices for Ceuta/Melilla, if that is your case
2021-06-15 08:30:13 ERROR (MainThread) [homeassistant.components.pvpc_hourly_pricing] Unique id '2.0TD' is already in use by 'sensor.pvpc'
Traceback (most recent call last):
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/components/pvpc_hourly_pricing/__init__.py", line 86, in async_setup_entry
    await async_migrate_entries(hass, entry.entry_id, update_unique_id)
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/helpers/entity_registry.py", line 744, in async_migrate_entries
    ent_reg.async_update_entity(entry.entity_id, **updates)
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/helpers/entity_registry.py", line 384, in async_update_entity
    return self._async_update_entity(
  File "/Volumes/WORK/contrib/home-assistant/homeassistant/helpers/entity_registry.py", line 457, in _async_update_entity
    raise ValueError(
ValueError: Unique id '2.0TD' is already in use by 'sensor.pvpc'
2021-06-15 08:30:13 ERROR (MainThread) [homeassistant.components.pvpc_hourly_pricing] Old PVPC-2 Sensor sensor.pvpc_extra is removed (another one already exists, using the same tariff)
```

which is the signal to remove the extra config entry with its extra sensor. This was the thing I had problems to reproduce on tests (I forgot to mock the entity-registry), but now it is solved and tested in `test_multi_sensor_migration`

About your comment about the log level for those:
> This is a warning, not an error.

I don't know the best way of doing this: I put an error there because the 'successful migration' message is already a warning ;-), as I wanted to point out to the user that the migrated sensor now can be configured with the options-flow, but I thought about generating a **persistent-notification** instead (is that a good idea?). In that case, success would generate the notification, and for 2nd and 3rd sensor, it would be just a warning (+ removal of the extra integration). I understand that errors should be used _only_ on non-desired/expected behaviours, but I didn't know how to differentiate those 2 possible outcomes (conversion ok vs removal of extra sensors). 

I could also reduce that to warning and finish with it, obviously, so whatever you consider best ;-)

Reply:
I think both log messages should be warning level. That's fine.

Reply:
Do we know which of the three existing entity entries should be removed before we start the migration? Then we could just remove the two that we don't want and then migrate the remaining one.

Reply:
> I think both log messages should be warning level. That's fine.

Ok, I'll change it

> Do we know which of the three existing entity entries should be removed before we start the migration? Then we could just remove the two that we don't want and then migrate the remaining one.

It doesn't matter. Neither of the 3 old tariffs have continuity into the new one (which is unique, except for a specific region of Spain, where it's another _variant_). That's why I did it so the first one survives and the rest, if any, are removed. And as said before, this is a very strange scenario (a user with >1 sensors), and, anyway, all old sensors working right now are showing now the same price (so they are in fact the same now. This change in the law affects all users of the old 3 tariffs, which now become just one

Reply:
If this case happens we should remove the surplus entity entries and finish the migration for the remaining entry. Currently, we remove one entry and then return False for config entry setup. That's not the correct solution.